### PR TITLE
Extractor2

### PR DESCRIPTION
**Pull request by Christian-B**
Required for Java and virtual machine branches
Needs to be merged again as somehow one commit disappeared!

includes
https://github.com/SpiNNakerManchester/SpiNNMachine/commit/e9abc92ff7697eb620ddf99fca1c5b14f7b5a6c1#diff-657d9411893649d855d59a4f0b61916a

### CODE REVIEW

File: spinn_machine/chip.py [hunk 1]
```diff
@@ -60,7 +60,9 @@ def __init__(self, x, y, processors, router, sdram, nearest_ethernet_x,
         :type ip_address: str
         :param virtual: boolean which defines if this chip is a virtual one
         :type virtual: bool
-        :param tag_ids: ID to identify the chip for SDP
+        :param tag_ids: IDs to identify the chip for SDP can be empty to
+            define no tags or None to allocate tag automatically
+            based on if there is an ip_address
         :type tag_ids: iterable(int)
```
Comment (alan-stokes) [inline spinn_machine/chip.py]: or none, otherwise pycharm will complain with incorrect python docs

File: spinn_machine/json_machine.py [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+            if "deadLinks" in details:
+                dead_links = details["deadLinks"]
+            else:
+                dead_links = []
+            links = []
+            for source_link_id in range(6):
+                if source_link_id not in dead_links:
+                    destination_x, destination_y = Machine.get_chip_over_link(
+                        source_x, source_y, source_link_id, width, height)
+                    opposite_link_id = (source_link_id + 3) % 6
```
Comment (alan-stokes) [inline spinn_machine/json_machine.py]: magic numbers

File: spinn_machine/json_machine.py [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+        :param machine: Machine to convert
+        :return: Map of Chip to list of virtual links
+        """
+        virtual_links_dict = defaultdict(list)
+        for chip in machine._virtual_chips:
+            # assume all links need special treatment
+            for link in chip.router.links:
+                virtual_links_dict[chip].append(link)
+                # Find and save inverse link as well
+                inverse_id = (
```
Comment (alan-stokes) [inline spinn_machine/json_machine.py]: this looks more like what i expected above

File: spinn_machine/json_machine.py [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+            for link in virtual_links_dict[chip]:
+                link_details = OrderedDict()
+                link_details["sourceLinkId"] = link.source_link_id
+                link_details["destinationX"] = link.destination_x
+                link_details["destinationY"] = link.destination_y
+                links.append(link_details)
+            details["links"] = links
+
+        exceptions = OrderedDict()
+        router_entries = JsonMachine._int_value(
```
Comment (alan-stokes) [inline spinn_machine/json_machine.py]: Ive read this code up above, thus cloned code. should be able to make that a method for both calls

File: spinn_machine/chip.py (hunk 1 above)
Comment (Christian-B) [reply spinn_machine/chip.py]: ok fiddle it

File: spinn_machine/json_machine.py (hunk 3 above)
Comment (alan-stokes) [reply spinn_machine/json_machine.py]: fixed

File: spinn_machine/json_machine.py (hunk 2 above)
Comment (alan-stokes) [reply spinn_machine/json_machine.py]: fixed

File: spinn_machine/json_machine.py (hunk 4 above)
Comment (alan-stokes) [reply spinn_machine/json_machine.py]: yea but no

**Pull request merged**